[ios] Added teardown of core objects when receiving UIApplicationWillTerminateNotification

julianrex opened the pull request:
Addresses #13120 (random crashes on app termination), by tearing down the underlying core C++ objects when receiving `UIApplicationWillTerminateNotification`.

This PR also saves the state of the `camera`, `debugMask` and `styleURL` during tear-down, so that they can be used when saving state (e.g. in `iosapp`
https://github.com/mapbox/mapbox-gl-native/blob/f4e5d842cbe15915a8854525ca928a914ba29eef/platform/ios/app/MBXViewController.m#L290)

In addition, this PR changes accesses to the underlying `mbgl::Map` to go via a method (`-(mbgl::Map &)[MGLMapView mbglMap]`) that raises an exception if the map is `NULL`.

This PR does *not* introduce `UIStateRestoring` conformance (see https://github.com/mapbox/mapbox-gl-native/issues/3660), but could be viewed as a prelude to that work.

friedbunny commented on platform/ios/src/MGLMapView.mm:
```diff
@@ -3091,29 +3150,29 @@ - (void)setZoomLevel:(double)zoomLevel animated:(BOOL)animated
 
-    _mbglMap->setZoom(zoomLevel,
+    self.mbglMap.setZoom(zoomLevel,
                       MGLEdgeInsetsFromNSEdgeInsets(self.contentInset),
```
Nit: update indentation.

friedbunny commented on platform/ios/src/MGLMapView.mm:
```diff
@@ -364,7 +368,11 @@ + (void)initialize
 - (nonnull NSURL *)styleURL
 {
-    NSString *styleURLString = @(_mbglMap->getStyle().getURL().c_str()).mgl_stringOrNilIfEmpty;
+    if (!_mbglMap) {
```
The brace style in `MGLMapView` is to place them on new lines:

```
if (blah)
{
    ...
}
```

julianrex replied:
Whoa. Ok, we have a few inconsistencies in the SDK. That's for another day ;)

1ec5 replied:
Yeah, the iOS-specific code started out with a brace-on-new-line style (as well as a couple other idiosyncrasies), but the rest of the Objective-C codebase (including platform/darwin/ and platform/macos/) generally uses a more Xcode-default style that puts braces on the same line as other code. Over time, that has caused the iOS-specific code to get a bit inconsistent. I personally try not to get worked up over whitespace, but some spring cleaning might eventually be needed. (PRs consistifying coding style tend to create lots of conflicts in outstanding PRs, so they can be a bit disruptive.)

julianrex commented on platform/ios/src/MGLMapView.mm:
```diff
@@ -1169,16 +1212,18 @@ - (void)willTerminate
     if (isVisible && ! _displayLink)
     {
-        if (_mbglMap->getConstrainMode() == mbgl::ConstrainMode::None)
+        if (_mbglMap && self.mbglMap.getConstrainMode() == mbgl::ConstrainMode::None)
```
I *think* we're safe here, but will double check.

julianrex replied:
This is only guarding the `setConstrainMode`, and `_mbglMap` should be valid from `init` through to termination. This is really me being paranoid.

1ec5 replied:
`mbglMap` is a C++ object. Messaging a nil Objective-C object is a no-op, but sending a message to a null C++ reference is undefined behavior and very likely to crash. If we expect the object to go away at some point during MGLMapView’s lifetime – and apparently we do – then the property should be backed by a pointer or perhaps an `optional<mbgl::Map>`.

julianrex commented on platform/ios/src/MGLMapView.mm:
```diff
@@ -3021,6 +3074,12 @@ - (void)_setCenterCoordinate:(CLLocationCoordinate2D)centerCoordinate animated:(
+        if (completion) {
+            completion();
+        }
+    }
```
Yes! Good catch

ChrisLoer commented:
Is the main thread message loop still running at the time `willTerminate` gets called? I'm wondering if all the checks to handle calling `MGLMapView` methods after it has destroyed the underlying map are actually necessary -- `willTerminate` itself shouldn't call back into any of those methods, right?

Also: I think in the same PR we should address other SDK objects that could be holding onto core objects that need teardown. The one I have in mind is `MGLMapSnapshotter` (which could also have worker threads running).

julianrex commented:
> I'm wondering if all the checks to handle calling MGLMapView methods after it has destroyed the underlying map are actually necessary -- willTerminate itself shouldn't call back into any of those methods, right?

The problem is that "other things" may be observing `UIApplicationWillTerminateNotification` and request properties on `MGLMapView`. In fact, that's exactly what happens with the current `iosapp` - it tries to save state, accessing `MGLMapView.camera`, which access the underlying (now gone) `mbgl::Map` and Boom.

My understanding is that at termination, the order goes something like:
1. Call application's `applicationWillTerminate:` method
2. Post notification (order of delivery is undefined)
3. exit

I changed the accesses to `_mbglMap` to go via a method (that raises an exception) to catch any other possible problems. This might be a game of whack-a-mole though.

> Also: I think in the same PR we should address other SDK objects that could be holding onto core objects that need teardown. The one I have in mind is MGLMapSnapshotter (which could also have worker threads running).

I'd like to get this PR in, so ideally this would be tail work. But you're right - I think we could have a bunch of things hanging on to core objects... (see also https://github.com/mapbox/mapbox-gl-native/pull/13356#discussion_r233506556)

ChrisLoer commented:
> Post notification (order of delivery is undefined)

Ugh yeah you're right. Mixing different lifecycle models is messy.

> I'd like to get this PR in, so ideally this would be tail work.

Fair enough. Just so it's clear, if a `MGLMapSnapshotter` is around, it will keep the _shared_ thread pool alive. Because the shared thread pool is still alive, the map object destructor _won't_ block waiting for threads to join, so it's still possible that the worker threads could be running when `exit` is called (even if the map snapshotter itself isn't doing any work -- the work could be from the main map object).

julianrex commented:
> Fair enough. Just so it's clear, if a MGLMapSnapshotter is around, it will keep the shared thread pool alive. Because the shared thread pool is still alive, the map object destructor won't block waiting for threads to join, so it's still possible that the worker threads could be running when exit is called (even if the map snapshotter itself isn't doing any work -- the work could be from the main map object).

Ok, we definitely need to address this more fully in a follow up PR. /cc @chloekraw @malwoodsantoro

1ec5 commented on platform/ios/src/MGLMapView.h:
```diff
@@ -636,11 +637,14 @@ MGL_EXPORT IB_DESIGNABLE
     location or `NO` if you want the map to display the new location
     immediately.
+ 
+ @note The behaviour of this method is undefined if called when observing
```
Nit: replace “behaviour” with “behavior”. (Sorry! 🇺🇸)

julianrex replied:
Weird - my Mac is actually set to US, so this must have been me forcing my 🇬🇧!

julianrex replied:
Will change.

1ec5 commented on platform/ios/src/MGLMapView.h:
```diff
@@ -653,6 +657,9 @@ MGL_EXPORT IB_DESIGNABLE
     zooming to the new location or `NO` if you want the map to display the new
     location immediately.
+ 
+ @note The behaviour of this method is undefined if called when observing
```
Nit: “when observing” sounds like it applies at any time after registering for the notification, which could be throughout the lifetime of the application. “In response to” would be clearer.

julianrex replied:
Will change.

1ec5 commented on platform/ios/src/MGLMapView.mm:
```diff
@@ -680,6 +696,45 @@ - (void)reachabilityChanged:(NSNotification *)notification
+    // Tear down C++ objects, insuring worker threads correctly terminate.
+    // Because of how _mbglMap is constructed, we need to destroy it first.
+    delete _mbglMap;
+    _mbglMap = nullptr;
```
`_mbglMap` is declared as an `mbgl::Map &`. A C++ reference can’t ever become null, because [a reference isn’t a pointer](https://stackoverflow.com/a/4364586/4585461). The Objective-C++ compiler apparently lets you get away with setting the reference to null, but using the reference is undefined behavior – even checking if the reference evaluates to `false`.

julianrex replied:
`_mbglMap` is still a `mbgl::Map *` - it's just the accessor that I changed to a C++ reference. This was to indicate that it should never be null, and if it is then a crash would be valid (hence raising the exception in the getter).

I'm in two minds about whether the getter should return a reference or pointer. Returning the pointer does indicate that care needs to be taken; OTOH we should have already taken care :)

julianrex commented:
Actually, using `optional<mbgl::Map>` might be the best solution here - thanks for reminding me it existed (my C++ knowledge is dated). I'll have a look at updating this PR.

julianrex commented on platform/ios/src/MGLMapView.mm:
```diff
@@ -680,6 +696,45 @@ - (void)reachabilityChanged:(NSNotification *)notification
+    _mbglThreadPool.reset();
+
+    _annotationContextsByAnnotationTag.clear();
+    _annotationTagsByAnnotation.clear();
```
> so I think we can be confident that mbgl worker threads would depend on them.

would or wouldn't?

1ec5 replied:
Wouldn’t. Good thing your 🔮 is working, because my logic checker isn’t. 😛

julianrex commented on platform/ios/src/MGLMapView.mm:
```diff
@@ -680,6 +696,45 @@ - (void)reachabilityChanged:(NSNotification *)notification
+    return @[@"camera", @"debugMask", @"styleURL"];
+}
+
+- (id)valueAtTerminationForKey:(NSString*)key ofClass:(Class)classType {
```
> #13120 is fixed by the other parts of this PR that explicitly check for an _mbglMap, correct?

Yes, though as you and @ChrisLoer say there are other parts of the code base that need tearing down, so perhaps it would be more accurate to say "mostly fixed".

julianrex commented:
> #13120 is fixed by the other parts of this PR that explicitly check for an _mbglMap, correct?

Yes, though as you and @ChrisLoer say there are other parts of the code base that need tearing down, so perhaps it would be more accurate to say "mostly fixed".

julianrex commented:
@ChrisLoer FYI I am going to look at `MGLMapSnapshotter` too

julianrex commented:
Rebased and updated. The latest changes don't touch `MGLOfflineStorage` (@ChrisLoer I'd love your input on whether this is strictly necessary), but does address `MGLMapSnapshotter`.

ChrisLoer commented:
> The latest changes don't touch MGLOfflineStorage

This seems a bit iffy to me, but `MGLOfflineStorage` itself doesn't hold onto a `ThreadPool`, which means in theory no one should be running its code during the `exit` (because the main thread's run loop has already shut down).

julianrex commented:
Rebased.

julianrex commented:
Thanks @friedbunny!
